Run CI Checks with a packaged version of fidesctl

ThomasLaPiana opened the pull request:
Closes #457 

### Code Changes

* [x] convert the Dockerfile to a multi-stage build
* [x] in one of the dockerfile build stages, do a non-symlinked install so we can catch packaging issues
* [x] update the CI checks to use the prod version of the image

### Steps to Confirm

* [x] tests should fail if we do something like remove an `__init__.py` file from a subdir and then run `make check-all`

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [x] Documentation Updated
* [x] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created

### Description Of Changes

To make sure that we're testing the same package as the one that's getting deployed, it was necessary to rework the Dockerfile into multiple stages so that we could have a nice `prod` image to upload that is the exactly same installation as would otherwise be coming from pypi (except it includes all of the deps).

Due to the complexity of the database adapters, it wasn't possible to shrink the image in a meaningful way unfortunately, but with the multi-stage builds already set up this would be easier to revisit in the future if needed.

ThomasLaPiana commented:
multi-stage image is 846MB
single-stage image is 1.33GB

ThomasLaPiana commented:
this needs to wait for #479 to get merged before being merged

ThomasLaPiana commented:
need to figure out some other ways to test this, maybe from Slack where people noted failures

ThomasLaPiana commented:
@ethyca/fides-control I need help thinking up some test cases here

PSalant726 commented:
I like @SteveDMurphy 's idea. Essentially what we're trying to achieve is confidence in our package's integrity. It makes sense to have a dedicated workflow that runs as part the the PR checks and as a deployment prerequisite, and simply attempts to build, install, and use fidesctl. We can also leverage a build matrix in the workflow to run this check for both PyPI and Conda (when we restore Conda), and for different system architectures.

ThomasLaPiana commented:
so for sure we are going to use this `prod` target to run our CI checks, but I'm not sure how to test the case where we want to make sure it breaks

ThomasLaPiana commented:
also @PSalant726 why is the `__init__.py` only now getting caught be xenon 😂

PSalant726 commented:
> so for sure we are going to use this `prod` target to run our CI checks, but I'm not sure how to test the case where we want to make sure it breaks

I'm not sure what you mean. In general, testing that something _doesn't_ happen (a successful build) is not a good practice, and is sometimes impossible. To me it seems like enough to test that things always build successfully, no?

> also @PSalant726 why is the `__init__.py` only now getting caught be xenon 😂

GAH 😡

ThomasLaPiana commented:
@PSalant726  agreed, i was getting lost in the testing sauce

I just wanted to do a double-check that if we did something like remove an `__init__.py` file that the test suite failed as we expect, but I can confirm that locally too

ThomasLaPiana commented:
### Testing:

1. added `RUN rm -rf src/fidesctl.egg-info` to the Dockerfile
2. deleted `include src/fidesctl/templates/fides_datamap_template.xlsx` from the `MANIFEST.in`
3. `make build-local-prod ; make pytest-unit` failed

1. removed `RUN rm -rf src/fidesctl.egg-info` from the Dockerfile
2. ran the same test as above, with the missing template file
3. the tests passed

This shows that the manifest is being respected and that the build isn't being affected by an existing `egg-info` directory. This wild issue was caused by the way that we rely on that same `egg-info` directory to develop our code live in the Docker container.

__NOTE__: I spent way too long tracking down this issue 👎 but we're better off for it

ThomasLaPiana commented on Makefile:
```diff
@@ -152,7 +156,7 @@ xenon:
 	--max-average A \
 	--ignore "data, tests, docs" \
-	--exclude "src/fidesctl/_version.py"
+	--exclude "src/fidesctl/_version.py,src/fidesctl/cli/__init__.py"
```
@PSalant726 im sorry, i just couldn't 🙃

sanders41 commented on fidesctl/Dockerfile:
```diff
@@ -1,8 +1,12 @@
+## Tool Installation ##
+#######################
 
 RUN apt-get update
```
A couple suggestions for your refactor.

By running `RUN apt-get update` on one line and `RUN apt-get install ...` on another, if you add new packages to install the `RUN apt-get update` layer will not run because it will be cached meaning you can miss security updates. By combining the update and install into the same lines you can avoid this. There are also a couple thing you can do here to reduce the overall size of the image.

```sh
 RUN : \
  && apt-get update \
  && apt-get install \
    -y --no-install-recommends \
    curl \
    git \
    ipython \
    make \
    vim \
    g++ \
    gnupg \
    gcc \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
```

sanders41 commented on fidesctl/Dockerfile:
```diff
@@ -15,9 +19,9 @@ RUN apt-get install -y \
+###################
 
 # Postgres
 RUN apt-get -y install libpq-dev
```
```sh
RUN : \
  && apt-get update \
  && apt-get install \
    -y --no-install-recommends \
    libpq-dev \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
```

sanders41 commented on fidesctl/Dockerfile:
```diff
@@ -32,26 +36,25 @@ RUN apt-get -y install \
     unixodbc-dev \
```
```sh
RUN curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add -
RUN curl https://packages.microsoft.com/config/debian/10/prod.list | tee /etc/apt/sources.list.d/msprod.list
ENV ACCEPT_EULA=y DEBIAN_FRONTEND=noninteractive
RUN : \
  && apt-get update \
  && apt-get install \
    -y --no-install-recommends \
    apt-transport-https \
    unixodbc-dev \
    mssql-tools \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
```

sanders41 commented on fidesctl/Dockerfile:
```diff
@@ -61,3 +64,26 @@ ENV RUNNING_IN_DOCKER=TRUE
+FROM builder as prod
+
+# Remove anything that could influence the build
+RUN rm -rf src/fidesctl.egg-info/
```
Could the `src/fidesctl.egg-info/`, `build/`, and `dist/` be added to the `.dockerignore` file and have the same result? If so that is a couple less steps for the build.

ThomasLaPiana replied:
they definitely could, i just put it here to be super sure

I'll move them to the ignore file and make sure it still works as intended

sanders41 reviewed (commented):
Just a couple small suggestions for you Dockerfile refactor.

ThomasLaPiana commented:
@sanders41 thanks for the comments!

PSalant726 commented on fidesctl/Dockerfile:
```diff
@@ -1,57 +1,74 @@
-RUN apt-get update
-RUN apt-get install -y \
+RUN : \
+    && apt-get update \
+    && apt-get install \
+    -y --no-install-recommends \
     curl \
     git \
     ipython \
     make \
     vim \
     g++ \
     gnupg \
-    gcc
+    gcc \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
```
Slick

PSalant726 commented on fidesctl/Dockerfile:
```diff
@@ -1,57 +1,74 @@
-## Database Dependencies ##
-###########################
+###################
+## Database Deps ##
```
A huge nit, and you don't need to change it, but I strongly prefer _not_ to abbreviate things in stored procedures like `Dockerfile`s, scripts, etc. We should be using the long-form wherever possible, since 1) these things aren't being typed out by a human and 2) it improves readability.

ThomasLaPiana replied:
I agree, clarity > brevity

PSalant726 commented on fidesctl/src/fidesctl/cli/__init__.py:
```diff
@@ -81,6 +81,8 @@ def cli(ctx: click.Context, config_path: str, local: bool) -> None:
 
     ctx.obj["CONFIG"] = config
+
+    # Run the help command if there isn't a subcommand
```
I don't believe this comment adds value. The conditional feels pretty straightforward to me.

PSalant726 reviewed (approved):
Some tiny nits. Nothing blocking 👍 

As an aside, should we be updating some of the Docker-related things in our other repos to align with the `apt-get`, `.dockerignore`, etc changes here?

ThomasLaPiana commented:
@PSalant726 good call

@SteveDMurphy i'd take note of this while you're working on the pypi stuff for fidescls (and preferably copy what we've done here)
